tests/Doctrine/Tests/ORM/Functional/HydrateAsCollectionTest.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional;
+
+use Doctrine\ORM\Tools\SchemaTool;
+use Doctrine\ORM\Query;
+use Doctrine\ORM\Internal\Hydration\ObjectHydrator;
+use Doctrine\ORM\Internal\Hydration\SimpleObjectHydrator;
+use Doctrine\Common\Collections\ArrayCollection;
+
+use Doctrine\Tests\Models\CMS\CmsUser;
+
+class HydrateAsCollectionTest extends \Doctrine\Tests\OrmFunctionalTestCase
+{
+    /**
+     * @var CustomEntityManager
+     */
+    protected $_cem;
+
+    protected function setUp()
+    {
+        $this->useModelSet('cms');
+        parent::setUp();
+        
+        $this->_cem = CustomEntityManager::fromEm($this->_em);
+    }
+
+    public function testLoadAsCollections()
+    {
+        $user = new CmsUser;
+        $user->name = 'Guilherme';
+        $user->username = 'gblanco';
+        $user->status = 'developer';
+        $this->_cem->persist($user);
+        $this->_cem->flush();
+        
+        $userId = $user->getId();
+        
+        $this->_cem->clear();
+        
+        $repos = $this->_cem->getRepository('Doctrine\Tests\Models\CMS\CmsUser');
+        
+        $user = $repos->find($userId);
+        $this->assertInstanceOf('Doctrine\Tests\Models\CMS\CmsUser', $user);
+
+        // Test for DDC-3987
+        $user = $repos->find(9001);
+        $this->assertNull($user);
+    }
+
+}
+
+/**
+ * Simulate an array collection class from other projects that
+ * has a different implementation on the ArrayAccess methods.
+ *
+ * {@inheritDoc}
+ */
+class SomeArrayCollection extends ArrayCollection
```

Comment:
It is a library that integrates doctrine into laravel. I do not understand what you mean with the collection would be replaced at hydration time. The authors of the library created a custom EntityManager that uses custom hydrators for Query::HYDRATE_OBJECT and Query::HYDRATE_SIMPLEOBJECT. These hydrators return instances of laravel's collection class instead of arrays. The custom hydrators actually look exactly the same as in the test. If that is an unwanted use-case to replace the standard hydrators with something that returns other data types you can reject the PR.